### ISSUE TITLE
ROS Gazebo black screen

### ISSUE THREAD

**Issue by IceSentry** [Microsoft/WSL#3368]
Following [this tutorial](https://janbernloehr.de/2017/06/10/ros-windows) I managed to run ros on wsl with ubuntu 16.04 and Microsoft Windows [Version 10.0.17134.112]. However when launching gazebo with VcXsrv it gives a black screen where you should see the 3d model of the robot, but the rest of the UI is there and there is no error in the console.

I know WSL is supposed to be for command line tools, but there have been a few issues related to using gui apps through x server that have been resolved here.

I created this issue after reading #1450 where it was suggested to create a new issue for this.

**Comment by Voskrese** [comment]
@IceSentry 
`vcxsrv.exe` :0 -ac -terminate -lesspointer -multiwindow -clipboard ...

`vcxsrv.exe :0 ` - display :0

**Comment by IceSentry** [comment]
@therealkenc it was asked on gazebo answer http://answers.gazebosim.org/question/16888/gazebo-on-wsl-bash-for-windows/ which is actually how I came to the original issue on github.

I did manage to see some gears with glxgears, unfortunately gazebo still shows a black screen. Although on my version of vcxsrv I don't see a native opengl checkbox. I use vcxsrv 1.20.0.0 from sourceforge

@Voskrese I have no idea what you want me to do

**Comment by shoffmeister** [comment]
glxgears gives me about 650 fps on vcxsrv 1.20 in native OpenGL mode on a comparatively slow Intel HD 4000 laptop GPU, about 1500 fps on an Intel HD 620 (two slow generations later)

You can enable OpenGL mode on the vcxsrv side by running XLaunch; this will take you through a wizard which on its last page refers to the LIBGL_ALWAYS_INDIRECT environment variable (`export LIBGL_ALWAYS_INDIRECT=1`)

Generally speaking, it would be nice to understand whether ROS gazebo actually does run with a *native* Linux and vcxsrv as the remote X server.

**Comment by IceSentry** [comment]
With export LIBGL_ALWAYS_INDIRECT=1 gazebo simply doesn't launch.

Also I made a mistake in my other comment, I was using an old version of xming. I tried it again with vcxsrv with native opengl and I had the same result.

**Comment by Voskrese** [comment]
@IceSentry  You wrote that you see a black screen, instead of anything, I assumed that you installed Xming, just do not set it on the desired display, by default-1, and you need 0, if of course you passed the same settings in UNIX environment

In general, you will be disassembled

**Comment by therealkenc** [comment]
>  this will take you through a wizard which on its last page refers to the LIBGL_ALWAYS_INDIRECT environment variable (export LIBGL_ALWAYS_INDIRECT=1)

Do not do that.

> Generally speaking, it would be nice to understand whether ROS gazebo actually does run with a native Linux and vcxsrv as the remote X server.

This is generally the prerequisite for posting here. Even then, monsters like ROS make for terrible use-cases for tracking down WSL bugs or feature gaps; since those bugs won't have to do with drawing pixels. An exception is being made since it was suggested to open an issue. 

> Although on my version of vcxsrv I don't see a native opengl checkbox. I use vcxsrv 1.20.0.0 from sourceforge

Start VcXsrv with XLaunch. You will see the checkbox on the third page of the wizard.

In any case, against my better judgement I installed ROS (a short part of my life I'll never get back). Worked here out of the box with Ubuntu 16.04 and VcXsrv 1.20. Which is pretty much what I feared.

n.b. I don't want to imply unchecking native gl is your magic bullet. I only know it lit up without incident for me. I am not sure the thing is actually _usable_, mind. I suspect Gazebo falls much into the same category as [Blender](https://github.com/Microsoft/WSL/issues/2764) on that front.

![image](https://user-images.githubusercontent.com/7049700/42478712-913273f8-838a-11e8-9cb6-c3af44711934.png)

**Comment by IceSentry** [comment]
I'm sorry if I wasted your time, I'm very new to all of this and since it was suggested in the other thread and I had a similar issue I made this thread.

I know WSL isn't supposed to be used for GUI apps, but I've seen a few issues here talking about running gui apps through an xserver so I assumed it was in a kind of grey area.

I will try to do a fresh wsl and gazebo install and see if it works. I'll also look into running gazebo on a 
a linux vm with an X server. To be clear I did find the native opengl option and I disabled it as suggested and it still didn't work for me.

But if I understand correctly you don't believe this is an issue with WSL and it is most likely with gazebo and/or the x server?

**Comment by therealkenc** [comment]
> I'm sorry if I wasted your time

Don't be. This isn't on you. It wasn't a waste. Yes it's a grey area.

> But if I understand correctly you don't believe this is an issue with WSL and it is most likely with gazebo and/or the x server?

Dunno, precisely, or I'd tell ya. I was kind of hoping `glxgears` _wouldn't_ work for _you_, in which case we could attack that head on. Or, that `gazebo` _wouldn't_ work for _me_, in which case I could deep dive for the curiosity value. We've got neither of those situations, unfortunately.

We do know "WSL" (namely the kernel emulation layer that makes this whole thing go) is fine. Because it works, demonstrably. WSL doesn't even know you are talking to VcXsrv. All WSL knows is you are writing some bytes down a socket on localhost port 6000.

It might be firewall related. I got quite a few popups to open ports. Try disabling your Windows firewall entirely and see if that helps. But we're grasping here.

You might try XMing just to switch things up. That at least purturbs a variable (the X Server). 

You can also try running it remotely from inside a VM as an experiment, if you are feeling motivated or are looking for a time sink. But trying a given scenario (like Gazebo) in a VM is more valuable in the case where you know it _doesn't_ work in WSL, and thus can start tracking down a diverge from the Real Thing. Here we know it does, so there is no diverge to discover. The possible outcomes of the experiment for you aren't pretty. Either it _doesn't_ work in the VM, which leaves you exactly where you are right now (wondering why). Or it _does_ work, which puts you in the unenviable position of trying to find out precisely what bits of your Real Linux userspace (Ubuntu + everything) differ from your WSL userspace (Ubuntu + everything). Which is "not easy", to put it mildly.

One "last" shot: what does `glxinfo | grep OpenGL` return for you. Should look like:
```
ken@DESKTOP-4UTIQSF:~/Devel/ros$ glxinfo | grep OpenGL
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 6.0, 256 bits)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 18.0.5
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 18.0.5
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 18.0.5
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
```
If it's the same just say 'same' no need to paste the output. I'm just trying to guess variables.

**Comment by shoffmeister** [comment]
On the subject of using `export LIBGL_ALWAYS_INDIRECT=1`, and also regarding "native windows opengl" in vcxsrv, I finally understand some of the implications there.

Fundamentally, using _native_ OpenGL forces all OpenGL commands to be rendered on the remote server (i.e. the "box" which eventually shows the bits and bytes). One of the side-effects of that is that the OpenGL command stream forces use of the GLX protocol - and that only supports OpenGL 1.4 (see http://x.cygwin.com/docs/ug/using-glx.html)

So, some experiments after `sudo apt install gazebo9`:
* launch `vcxsrv` _with_ "Native opengl" checked
-- `export LIBGL_ALWAYS_INDIRECT=1 && gazebo` -> exactly nothing happens, silent crash (gazebo diagnostics probably broken)
-- `export LIBGL_ALWAYS_INDIRECT=0 && gazebo` ->
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
```

* launch `vcxsrv` _without_ "Native opengl" checked
-- `export LIBGL_ALWAYS_INDIRECT=1 && gazebo` -> exactly nothing happens, silent crash (gazebo diagnostics probably broken)
-- `export LIBGL_ALWAYS_INDIRECT=0 && gazebo` -> **life is good**

At no time am I able to see a black screen; something must be seriously broken elsewhere for that to happen (in case "black screen" is exactly what is visible)

**Comment by shoffmeister** [comment]
> But if I understand correctly you don't believe this is an issue with WSL and it is most likely with gazebo and/or the x server?

The probability of this being a WSL problem is very, very low, given the above.

**Comment by IceSentry** [comment]
![image](https://user-images.githubusercontent.com/8348954/42531420-ea5a8cf0-8451-11e8-8ec0-4263569180e8.png)

This is what I mean by black screen I should have made this at the beginning. The issue is that the 3d viewer is black not the entire window.

I'll do a clean install of wsl, gazebo and vcxsrv tomorrow and try to investigate more.

**Comment by yxy1996** [comment]
I have the same black screen problem, with a "preparing world" window on the front.

**Comment by therealkenc** [comment]
That warning is a window manager thing (xfce4) and won't be related. [Also, running a window manager serves no purpose, but I digress.]

I think this one has about run its course. There are [reports](https://www.google.ca/search?q=gazebo+black+screen&sa=X&ved=0ahUKEwjh1ZeQ1rrcAhXNJDQIHZQOD3gQ1QIIeygB&biw=1692&bih=916) of the [same behaviour](https://bitbucket.org/osrf/gazebo/issues/2457/gazebo-black-screen-and-no-world-model) on Real Linux. Try disabling the Windows firewall (there was no response to that [suggestion](https://github.com/Microsoft/WSL/issues/3368#issuecomment-403709867)), but that is mostly pissing on a [spark plug](https://www.youtube.com/watch?v=7FogV4Qv8p8). If `glxgears` was _also_ drawing blackness we could explore _general_ remote opengl problems, but that isn't the case here. The way forward here would be for someone motivated to deep dive Gazebo and find out exactly what it is up to. But there's no WSL actionable, and no one here purports to be a Gazebo expert (I had never heard of the thing before this issue was submitted). Bonne chance.

**Comment by jbohren-hbr** [comment]
@therealkenc thanks for spending time on this. I encountered the same issue that  @IceSentry is seeing, and there's an easy solution:

```
export GAZEBO_IP=127.0.0.1
```

---

I tried running this out of the box, and also saw a black render window, so I started with your suggestion:

> Try disabling the Windows firewall (there was no response to that suggestion),

With the firewall completely disabled, I still saw a black render window when following the normal gazebo bringup. This did appear to be related to the network configuration, though, as the gazebo server emits warnings about outgoing message queues filling up (nobody is receiving them). I also had the same issue when running as admin.

Below is the repro that shows and then fixes the bug:

---

From vanilla Windows 10 Pro x64 with Intel graphics:
1. Install WSL [ref](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
```
win$ Enable-WindowsOptionalFeature -Online -FeatureName Microsoft-Windows-Subsystem-Linux
```
2. Install Ubuntu16.04 from the MS Store [link](https://www.microsoft.com/en-us/p/ubuntu-1604/9pjn388hp8c9?activetab=pivot%3aoverviewtab)
3. Install VcXsrv 64.1.20.1.0 [link](https://sourceforge.net/projects/vcxsrv/)
4. Launch VcXsrv through XLaunch with the following commands:
![image](https://user-images.githubusercontent.com/25538748/44409328-302bf300-a530-11e8-8698-518839c31dce.png)
![image](https://user-images.githubusercontent.com/25538748/44409343-35893d80-a530-11e8-8c88-fa3b687ec5b4.png)
![image](https://user-images.githubusercontent.com/25538748/44409352-391cc480-a530-11e8-95c3-75c16c13b6ef.png)
5. Launch Ubuntu16.04 in WSL (and initialize with new user creds)
6. Install `xubuntu-desktop` to suppress window manager warnings
```bash
linux$ sudo apt install xubuntu-desktop
```
7. Install `mesa-utils` to verify 3d acceleration
```bash
linux$ sudo apt install mesa-utils`
```
8. Set `DISPLAY`
```bash
linux$ export DISPLAY=:0
```
9. Run glxgears to verify 3d acceleration:
```bash
linux$ glxgears
```
10. Install gazebo 9 with bootstrap script:
```bash
linux$ curl -sSL http://get.gazebosim.org | sh
```
11. Run gazebo:
```bash
linux$ gazebo --verbose
```

Result:
![image](https://user-images.githubusercontent.com/25538748/44409891-3ff80700-a531-11e8-9b74-0dee8a275994.png)

```
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 169.254.241.102
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 169.254.241.102
```

***What stands out here is the `[Msg]  Publicized address: 169.254.241.102` since that is not the ip address for any valid interface according to `ifconfig`.***

So, killing gazebo and exporting `GAZEBO_IP` [ref](http://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/group__gazebo__transport.html) fixes this:
```
linux$ export GAZEBO_IP=127.0.0.1
linux$ gazebo --verbose
```

![working](https://user-images.githubusercontent.com/25538748/44411421-77b47e00-a534-11e8-8016-3322b687a4ed.png)

```
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 127.0.0.1
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 127.0.0.1
```

I still see a lot of message errors, but this is unlikely a WSL-specific issue beyond networking config:
```
Exception sending a message
...
```

@therealkenc, as you said, it runs, but usability is still yet to be seen.

Shoutout @nkoenig @scpeters @tfoote

**Comment by yxy1996** [comment]
OMG! I have had just tried this, and it really worked! In fact I launched it succeessfully before, with waiting for it about 20min, which made me thoroughly and finally give up. However, using your solution to launch seems much more normal and save me another 19min.

THANKS! A LOT!
发件人: Jonathan Bohren
发送时间: 2018年8月21日 23:33
收件人: Microsoft/WSL
抄送: yxy1996; Comment
主题: Re: [Microsoft/WSL] ROS Gazebo black screen (#3368)

@therealkenc thanks for spending time on this. I encountered the same issue that @IceSentry is seeing, and there's an easy solution:
export GAZEBO_IP=127.0.0.1

I tried running this out of the box, and also saw a black render window, so I started with your suggestion:
Try disabling the Windows firewall (there was no response to that suggestion),
With the firewall completely disabled, I still saw a black render window when following the normal gazebo bringup. This did appear to be related to the network configuration, though, as the gazebo server emits warnings about outgoing message queues filling up (nobody is receiving them). I also had the same issue when running as admin.
Below is the repro that shows and then fixes the bug:

From vanilla Windows 10 Pro x64 with Intel graphics:
1. Install WSL ref
win$ Enable-WindowsOptionalFeature -Online -FeatureName Microsoft-Windows-Subsystem-Linux
2. Install Ubuntu16.04 from the MS Store link
3. Install VcXsrv 64.1.20.1.0 link
4. Launch VcXsrv through XLaunch with the following commands:



5. Launch Ubuntu16.04 in WSL (and initialize with new user creds)
6. Install xubuntu-desktop to suppress window manager warnings
linux$ sudo apt install xubuntu-desktop
7. Install mesa-utils to verify 3d acceleration
linux$ sudo apt install mesa-utils`
8. Set DISPLAY
linux$ export DISPLAY=:0
9. Run glxgears to verify 3d acceleration:
linux$ glxgears
10. Install gazebo 9 with bootstrap script:
linux$ curl -sSL http://get.gazebosim.org | sh
11. Run gazebo:
linux$ gazebo --verbose
Result:

Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 169.254.241.102
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 169.254.241.102
What stands out here is the [Msg] Publicized address: 169.254.241.102 since that is not the ip address for any valid interface according to ifconfig.
So, killing gazebo and exporting GAZEBO_IP ref fixes this:
linux$ export GAZEBO_IP=127.0.0.1
linux$ gazebo --verbose

Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
Gazebo multi-robot simulator, version 9.3.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 127.0.0.1
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 127.0.0.1
I still see a lot of message errors, but this is unlikely a WSL-specific issue beyond networking config:
Exception sending a message
...
@therealkenc, as you said, it runs, but usability is still yet to be seen.
Shoutout @nkoenig @scpeters @tfoote

**Comment by therealkenc** [comment]
Glad you got it working. 👍

> 8. Set DISPLAY
> `linux$ export DISPLAY=:0`

Minor observation, setting aside `$GAZEBO_IP` for a moment. Your `$DISPLAY` variable wants to be `127.0.0.1:0` or `localhost:0` as well (per [earlier](https://github.com/Microsoft/WSL/issues/3368#issuecomment-403400778)). It might not (and probably doesn't) have any effect on Gazebo, but "`:0`" means X11 over `AF_UNIX` on `/tmp/.X11-unix/X0`, not `AF_INET` over `localhost:6000`. But this can (and [does](https://github.com/Microsoft/WSL/issues/648#issuecomment-284855564)) cause problems in some scenarios. 

Mentioning it mostly because I _didn't_ personally set `$GAZEBO_IP` _despite_ it working anyway. That difference remains somewhat unexplained. I did however blindly install all of ROS (because the OP said do that), which pulls in the kitchen sink. So who knows what variables got set. Either way, nice follow up.

**Comment by jbohren-hbr** [comment]
> Your `$DISPLAY` variable wants to be `127.0.0.1:0` or `localhost:0` as well (per earlier). It might not (and probably doesn't) have any effect on Gazebo, but `":0"` means X11 over `AF_UNIX` on `/tmp/.X11-unix/X0`, not `AF_INET` over `localhost:6000`. But this can (and does) cause problems in some scenarios.

Ah, that's good clarification.

> Mentioning it mostly because I didn't personally set `$GAZEBO_IP` despite it working anyway. That difference remains somewhat unexplained.

Gazebo looks like it's using `getifaddrs` [(location)](https://bitbucket.org/osrf/gazebo/src/f4e51e8f64001beda70615ab8982a7d5a4e65b18/gazebo/transport/Connection.cc?at=default&fileviewer=file-view-default#Connection.cc-737:744) [(manpage)](http://man7.org/linux/man-pages/man3/getifaddrs.3.html) to determine the IP of a given host. WSL networking is also something I lack knowledge about, too, but since that's in the linux kernel, this _might actually_ be a WSL bug. I see that there are some previous issues with `getifaddrs`: https://github.com/Microsoft/WSL/issues/308

> I did however blindly install all of ROS (because the OP said do that), which pulls in the kitchen sink. So who knows what variables got set.

Yeah, I already had a ROS installation in my WSL (but not the machine I used for vanilla repro/debug), and it still failed in the same way.  I don't think any package in `gazebo_ros_pkgs` exports `GAZEBO_IP`, anyway.

**Comment by jbohren-hbr** [comment]
It looks like instead of coming back as `AF_INET` like it does on linux, on WSL, `sa_family` for every interface is `AF_PACKET` [(manpage)](http://man7.org/linux/man-pages/man7/packet.7.html). 
@therealkenc

**Comment by therealkenc** [comment]
> It's still not clear why getifaddrs is returning this additional 169.254.x.x IP address on WSL and not native Linux, though.

That's excellent information to add, thanks. Since it did "work" for me out of the box (possibly due to differences in my network rig) and, more practically, because I've long since wiped ROS/Gazebo, I ~will need a high bar of boredom to~ probably won't deep dive Gazebo as the repro of choice for a `getifaddrs` diverge. But absolutely, consider your observation well noted (for when not if something else comes up that smells like a `getifaddrs` diverge). If you can do a small tight CLI test case you'd definitely get a shiny gold star, but the challenge is going to be doing one concise enough to actually reproduce for the folks in Redmond (read: "returns this additional 169.254.x.x IP").

**Comment by jbohren-hbr** [comment]
> If you can do a small tight CLI test case you'd definitely get a shiny gold star, but the challenge is going to be doing one concise enough to actually reproduce for the folks in Redmond (read: "returns this additional 169.254.x.x IP").

Here you go (no shiny stars needed):
https://github.com/jbohren-hbr/ifaddrs_test

This works on my native xubuntu 16.04 machine, fails on WSL with extra phantom interfaces that don't show up in `ifconfig`  and are all on the `169.254.x.x` subnet, which is not the right subnet for any of the interfaces I actually have:

```
wsl$ ./test.bash
not matching.
1,2d0
< eth0
< eth1
7,8d4
< eth6
< eth7
11,12d6
< wifi1
< wifi2
```

```
native$ ./test.bash
matching.
```

**Comment by therealkenc** [comment]
Okay. Unfortunately (?) your test doesn't repro for me here. I get "`matching.`" on WSL 18219. Which is, of course, why Gazebo worked out of the box for me (I would have been on some earlier Insiders in July).

I can't think of a dupe off the cuff that would explain a fix since 17134 (usually I can if one exists, but that doesn't mean it doesn't exist). I don't have a 17134 handy to try. First step would be to try Insiders (fast ring not skip-ahead will do) and see if it reproduces there.

Note also your `./text.bash` script is sensitive to the `ifconfig` output syntax, which appears to have changed between Ubuntu 16.04 and 18.04 (`ifconfig | grep -B1 "inet addr"` won't get a hit on Real Linux Ubuntu 18.04). Not a problem, but something to be aware of since we've got Windows version number **and** Ubuntu version number variables at play here.

**Comment by jbohren-hbr** [comment]
OK. I don't think this is an issue with WSL. The issue seems to be that `getifaddrs` returns interfaces that are both up and down. Gazebo should be checking if `IFF_UP` in the flags, surprise.

See here: https://github.com/jbohren-hbr/ifaddrs_test/blob/master/print_ifaddrs.cpp#L92

Test now passing, this can be put to rest. Need to create a PR on gazebo though...

**Comment by jbohren-hbr** [comment]
For completeness, Gazebo PR with fix is here: https://bitbucket.org/osrf/gazebo/pull-requests/3009/adding-check-to-make-sure-automatic/diff

**Comment by therealkenc** [comment]
🌟

**Comment by dantistus** [comment]
I have encountered same issue on Ubuntu Bionic, ROS Melodic and Gazebo 9 (running on virtual machine). I have 'solved' it by killing previously previously launched instance of gzserver.

**Comment by karray** [comment]
> I still see a lot of message errors, but this is unlikely a WSL-specific issue beyond networking config:
> 
> ```
> Exception sending a message
> ...
> ```

The same here. I've installed ROS-Melodic (full desktop) with Gazebo 9 on 3 different machines: 2 of them are with Ubuntu 18.04 and third one with WSL Ubuntu 18.04 (Windows 10). After a fresh install I get `Exception sending a message`, which is being printed infinitely after starting gazebo (gzserver, roslaunch).  But it seems that everything.

I've also noticed that it happens only when I open **VPN connection**

**Comment by traversaro** [comment]
It seems that a similar issue related to `169.245.*.*` "phantoom" network interfaces is affecting also the Windows build of Gazebo, see https://github.com/ms-iot/ROSOnWindows/issues/47#issuecomment-450665704 . This may indicate that the problem/feature of phantom interfaces may not be related to WSL at all, but it could be related just to the Windows network stack.

**Comment by traversaro** [comment]
If I understand correctly (see https://github.com/ms-iot/ROSOnWindows/issues/47#issuecomment-450668204) the `169.245.*.*` interfaces are related to ädapter that are disconnected, and so the issues is in Gazebo that should ignore them.

**Comment by therealkenc** [comment]
> so the issues is in Gazebo

You are now caught up to [above](https://github.com/Microsoft/WSL/issues/3368#issuecomment-415278072).

**Comment by traversaro** [comment]
> You are now caught up to above.

Thanks! Sorry for missing that comment.

**Comment by achimismaili** [comment]
For me, setting DISPLAY to 0, 0.0, localhost:0 or 127.0.0.1 on a ubuntu windows linux sub system (WLS2) did all not work, I had to use the public IP and the following command helped me:

`export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0`

**Comment by luyor** [comment]
I use ROS Gazebo in WSL docker. For me, the network settings are correct, but I still see the blank screen. The reason is, Gazebo is downloading data silently in background, and I just need to wait for it to finish.

I use this [tutorial ](https://marinerobotics.gtorg.gatech.edu/running-ros-with-gui-in-docker-using-windows-subsystem-for-linux-2-wsl2/) to setup ROS Gazebo in WSL docker. 
I run `roscore` in terminal. Then run `rosrun gazebo_ros gazebo` in second terminal. I have a blank screen as @IceSentry, and a log in terminal: `waitForService: Service [/gazebo/set_physics_properties] has not been advertised, waiting...`

Gazebo is downloading data silently in background, see discussion in the [forum](https://answers.ros.org/question/264473/gazeboset_physics_properties-question-be-resolved/).
Close the window, and wait for the log to print in terminal:
```
[ INFO] [1613558941.011174200]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[ INFO] [1613558941.208707300]: Physics dynamic reconfigure ready.
```

Rerun `rosrun gazebo_ros gazebo`, then it works.

**Comment by cmitch419** [comment]
> @therealkenc thanks for spending time on this. I encountered the same issue that @IceSentry is seeing, and there's an easy solution:
> 
> ```
> export GAZEBO_IP=127.0.0.1
> ```
> 
> I tried running this out of the box, and also saw a black render window, so I started with your suggestion:
> 
> > Try disabling the Windows firewall (there was no response to that suggestion),
> 
> With the firewall completely disabled, I still saw a black render window when following the normal gazebo bringup. This did appear to be related to the network configuration, though, as the gazebo server emits warnings about outgoing message queues filling up (nobody is receiving them). I also had the same issue when running as admin.
> 
> Below is the repro that shows and then fixes the bug:
> 
> From vanilla Windows 10 Pro x64 with Intel graphics:
> 
> 1. Install WSL [ref](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
> 
> ```
> win$ Enable-WindowsOptionalFeature -Online -FeatureName Microsoft-Windows-Subsystem-Linux
> ```
> 
> 1. Install Ubuntu16.04 from the MS Store [link](https://www.microsoft.com/en-us/p/ubuntu-1604/9pjn388hp8c9?activetab=pivot%3aoverviewtab)
> 2. Install VcXsrv 64.1.20.1.0 [link](https://sourceforge.net/projects/vcxsrv/)
> 3. Launch VcXsrv through XLaunch with the following commands:
>    ![image](https://user-images.githubusercontent.com/25538748/44409328-302bf300-a530-11e8-8698-518839c31dce.png)
>    ![image](https://user-images.githubusercontent.com/25538748/44409343-35893d80-a530-11e8-8c88-fa3b687ec5b4.png)
>    ![image](https://user-images.githubusercontent.com/25538748/44409352-391cc480-a530-11e8-95c3-75c16c13b6ef.png)
> 4. Launch Ubuntu16.04 in WSL (and initialize with new user creds)
> 5. Install `xubuntu-desktop` to suppress window manager warnings
> 
> ```shell
> linux$ sudo apt install xubuntu-desktop
> ```
> 
> 1. Install `mesa-utils` to verify 3d acceleration
> 
> ```shell
> linux$ sudo apt install mesa-utils`
> ```
> 
> 1. Set `DISPLAY`
> 
> ```shell
> linux$ export DISPLAY=:0
> ```
> 
> 1. Run glxgears to verify 3d acceleration:
> 
> ```shell
> linux$ glxgears
> ```
> 
> 1. Install gazebo 9 with bootstrap script:
> 
> ```shell
> linux$ curl -sSL http://get.gazebosim.org | sh
> ```
> 
> 1. Run gazebo:
> 
> ```shell
> linux$ gazebo --verbose
> ```
> 
> Result:
> ![image](https://user-images.githubusercontent.com/25538748/44409891-3ff80700-a531-11e8-9b74-0dee8a275994.png)
> 
> ```
> Gazebo multi-robot simulator, version 9.3.1
> Copyright (C) 2012 Open Source Robotics Foundation.
> Released under the Apache 2 License.
> http://gazebosim.org
> 
> [Msg] Waiting for master.
> [Msg] Connected to gazebo master @ http://127.0.0.1:11345
> [Msg] Publicized address: 169.254.241.102
> Gazebo multi-robot simulator, version 9.3.1
> Copyright (C) 2012 Open Source Robotics Foundation.
> Released under the Apache 2 License.
> http://gazebosim.org
> 
> [Msg] Waiting for master.
> [Msg] Connected to gazebo master @ http://127.0.0.1:11345
> [Msg] Publicized address: 169.254.241.102
> ```
> 
> _**What stands out here is the `[Msg] Publicized address: 169.254.241.102` since that is not the ip address for any valid interface according to `ifconfig`.**_
> 
> So, killing gazebo and exporting `GAZEBO_IP` [ref](http://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/group__gazebo__transport.html) fixes this:
> 
> ```
> linux$ export GAZEBO_IP=127.0.0.1
> linux$ gazebo --verbose
> ```
> 
> ![working](https://user-images.githubusercontent.com/25538748/44411421-77b47e00-a534-11e8-8016-3322b687a4ed.png)
> 
> ```
> Gazebo multi-robot simulator, version 9.3.1
> Copyright (C) 2012 Open Source Robotics Foundation.
> Released under the Apache 2 License.
> http://gazebosim.org
> 
> [Msg] Waiting for master.
> Gazebo multi-robot simulator, version 9.3.1
> Copyright (C) 2012 Open Source Robotics Foundation.
> Released under the Apache 2 License.
> http://gazebosim.org
> 
> [Msg] Connected to gazebo master @ http://127.0.0.1:11345
> [Msg] Publicized address: 127.0.0.1
> [Msg] Waiting for master.
> [Msg] Connected to gazebo master @ http://127.0.0.1:11345
> [Msg] Publicized address: 127.0.0.1
> ```
> 
> I still see a lot of message errors, but this is unlikely a WSL-specific issue beyond networking config:
> 
> ```
> Exception sending a message
> ...
> ```
> 
> @therealkenc, as you said, it runs, but usability is still yet to be seen.
> 
> Shoutout @nkoenig @scpeters @tfoote

You're a saint. This fixed the issue for me, thank you!

**Comment by besch** [comment]
For me the 'black screen in gazebo' fix was to install latest Intel [driver](https://www.intel.com/content/www/us/en/download/19344/intel-graphics-windows-dch-drivers.html
)
Although one issue left to solve is, after Windows restart gazebo scene resets to black screen

**Comment by lidianzhong** [comment]
It took me a long time to finally solve the problem by deleting the `.wslconfig` content. The problem should be related to the ip network problem.
![image](https://github.com/microsoft/WSL/assets/93401265/909857ed-41e3-4b68-8398-e1134b759c13)

**Comment by TomE8** [comment]
I had the same problem:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/077aa67c-24c6-4268-b958-e6bca6597c08" />

The solution for me was to use ogre engine:
`gz sim --render-engine ogre`

For the record, I run Gazebo on docker inside a virtual box.

**Comment by chidalu-z17** [comment]
I had the same problem but mine was a bit weird as I had gazebo installed in a docker where my ros2-humble was installed, and that docker was installed on my oracle vm virtual box which was installed on my windows 11 PC.
Essentially Gazebo -> ROS2-Humble Container (Docker based on Ubuntu 22.04) -> Ubuntu 24.04 -> Oracle VM VirtualBox -> Windows 11 PC.

I think what worked for me was to:
run: xhost +local:root or alternativelty run: xhost +local:docker (would work too and its generally safer than using root) on my ubuntu 24.04 terminal (host) and I increased the number of CPU allocated to the Ubuntu 24.04 VM to 3, instead of the default 2, as Gazebo requires enough recourses to render in 3D.

It could also had been that Gazebo was "loading" and increasing the number of CPU's essentially made it load faster. I haven't confirmed this. You could try waiting on the blank screen to go away too. If not, then just try what I described above.
